### PR TITLE
Add issuer_name and reason_code to X509RevokedRef

### PR DESCRIPTION
**Pull request by Skepfyr**

### CODE REVIEW

File: openssl/src/x509/mod.rs [hunk 1]
```diff
@@ -1421,6 +1422,37 @@ impl X509ReqRef {
     }
 }
 
+/// The reason that a certificate was revoked.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct CrlReason(i64);
```
Comment (Skepfyr) [inline openssl/src/x509/mod.rs:1427]: This feels messy but the ASN1_ENUMERATED `get_int64` methods forces my hand slightly.

File: openssl/src/asn1.rs [hunk 2]
```diff
@@ -661,6 +661,32 @@ cfg_if! {
...
+
+    /// An ASN.1 enumerated.
+    pub struct Asn1Enumerated;
+    /// A reference to an [`Asn1Enumerated`].
+    pub struct Asn1EnumeratedRef;
+}
+
+impl Asn1EnumeratedRef {
+    /// Get the value, if it fits in the required bounds.
+    #[corresponds(ASN1_ENUMERATED_get)]
```
Comment (sfackler) [inline openssl/src/asn1.rs:676]: ASN1_ENUMERATED_get_int64

File: openssl/src/x509/mod.rs [hunk 3]
```diff
@@ -1472,6 +1511,46 @@ impl X509RevokedRef {
...
+                self.as_ptr() as *const _,
+                // NID_crl_reason is an X509_REVOKED extension that is an ASN1_ENUMERATED
+                ffi::NID_crl_reason,
+                // Only one instance of the extension is permissable
+                ptr::null_mut(),
+                // Don't care if the extension is critical
+                ptr::null_mut(),
+            ) as *mut _)
+        }?;
+        Some(reason_code.get_i64().map(CrlReason::from_raw))
```
Comment (sfackler) [inline openssl/src/x509/mod.rs:1552]: I'm not sure how I feel about this doing the conversion to CrlReason internally.

File: openssl/src/x509/mod.rs [hunk 4]
```diff
@@ -1472,6 +1511,46 @@ impl X509RevokedRef {
             Asn1IntegerRef::from_ptr(r as *mut _)
         }
     }
+
+    /// Get the issuer name of the revoked certificate
+    #[corresponds(X509_REVOKED_get_ext_d2i)]
+    pub fn issuer_name(&self) -> Option<Stack<GeneralName>> {
```
Comment (sfackler) [inline openssl/src/x509/mod.rs:1517]: In keeping with the general attempt to have 1:1 bindings to OpenSSL, how about switching to a single generic method that retrieves extensions? Something like:

```rust
impl X509RevokedRef {
    pub fn extension<T>(&self) -> Option<T::Output>
        where T: X509Extension
    { ... }
}

pub unsafe trait X509Extension {
    const NID: Nid;

    type Output: ForeignType;
}

pub enum CertificateIssuer {}

unsafe impl X509Extension for CertificateIssuer {
    const NID: Nid = Nid::CERTIFICATE_ISSUER;

    type Output = Stack<GeneralName>;
}
```

File: openssl/src/x509/mod.rs (hunk 1 above)
Comment (Skepfyr) [reply openssl/src/x509/mod.rs]: I've changed this to use `c_int` as it's now the caller's problem.

File: openssl/src/x509/mod.rs (hunk 3 above)
Comment (Skepfyr) [reply openssl/src/x509/mod.rs]: I like it as it feels a little more type-safe with no downsides I can see (CRLReason has a from_raw and to_raw). It's moot now though as the trait based system can't do it.

File: openssl/src/x509/mod.rs (hunk 4 above)
Comment (Skepfyr) [reply openssl/src/x509/mod.rs]: I've swapped over to this, I think it's a nicer match for the OpenSSL API and we should probably do the same thing to all the others but it is definitely a bit more complicated. (I did also deal with the crit parameter correctly at the same time which didn't help).

File: openssl/src/x509/mod.rs [hunk 5]
```diff
@@ -50,6 +50,15 @@ pub mod store;
 #[cfg(test)]
 mod tests;
 
+/// A type of X509 extension.
+///
+/// # Safety
+/// The value of NID and Output must match those in OpenSSL so that
```
Comment (sfackler) [inline openssl/src/x509/mod.rs:56]: Docs got cut off.

**Comment by sfackler** [comment]
Nice - looks good! Can you add a few quick tests just confirming that the extension impls added work as expected?

File: openssl/src/x509/mod.rs (hunk 5 above)
Comment (Skepfyr) [reply openssl/src/x509/mod.rs:56]: Whoops, thanks, fixed.

**Pull request merged**